### PR TITLE
added iterative imputer

### LINKED ISSUE #654: [OpenSTEF 4.0] Add support for IterativeImputer for feature imputation
https://scikit-learn.org/stable/modules/generated/sklearn.impute.IterativeImputer.html
Idea is that when data is missing imputer can impute the data with a smaller model.

> [!IMPORTANT]
> Changes should be done by forking [OpenSTEF 4.0.0 tracking branch](https://github.com/OpenSTEF/openstef/tree/release/v4.0.0) (`release/v4.0.0`) and then make a PR back onto `release/v4.0.0` branch.
> Please read the [contributing guide](https://openstef.github.io/openstef/v4/contribute/index.html) before you start.

Definition of Done:

✅ Acceptance criteria

- [ ] Used existing imputer (https://scikit-learn.org/stable/modules/generated/sklearn.impute.IterativeImputer.html)
- [ ] Implemented as TimeSeriesTransform or as en extra case for existing ImputerTransform
- [ ] User can pick which columns / features to impute and the relevant covariants.
- [ ] Edge cases are documented
- [ ] Unit tests are created
- [ ] Docstrings and lint checks pass.

📄 Documentation criteria:

Interfaces with other products and services

Update relevant confluence pages

🧪 Test criteria:

Unit tests written and passed

⌛ Dependencies:

N/A

🚀 Releasing:

N/A

Other information:

🌍 Background

### PR DESCRIPTION
**Pull request by fleur-petit**
closes #654 

- Added strategy "iterative" to Imputer transform.
  - The attribute "impute_estimator" can be used to pass desired estimator (for example BayesianRidge (default), or RandomForestRegressor)
- Tests added:
  - parametrise basic test to test different strategies
  - test if iterative imputer can find linear relationships between features
  - test should work better than mean if missing value is not coincidentally the mean
  - test `initial_strategy` is used when only one feature is given with iterative strategy 

✅ Acceptance criteria

- [x] Used existing imputer (https://scikit-learn.org/stable/modules/generated/sklearn.impute.IterativeImputer.html)
- [x] Implemented as extra case for existing Imputer
- [x] User can pick which columns / features to impute and the relevant covariants.
- [x] Edge cases are documented
- [x] Unit tests are created
- [x] Docstrings and lint checks pass.

### CODE REVIEW

File: packages/openstef-models/src/openstef_models/transforms/general/imputer.py [hunk 1]
```diff
@@ -112,12 +150,31 @@ class Imputer(BaseConfig, TimeSeriesTransform):
...
+        description="Tolerance for IterativeImputer convergence",
+    )
+    max_iterations: int = Field(
+        default=40,
+        description="Maximum iterations for IterativeImputer",
+    )
     selection: FeatureSelection = Field(
         default=FeatureSelection.ALL,
-        description="Features to impute.",
+        description=(
```
Comment (fleur-petit) [inline packages/openstef-models/src/openstef_models/transforms/general/imputer.py:171]: @bartpleiter is the description of the `selection` for "iterative" strategy clear enough? Is it not too confusing that except for columns for imputation, a user can also select predictor columns here? The the distinction between columns to impute and column to use as predictor is not so clear in IterativeImputer: all selected columns are both used as predictors as well as imputed when they have missing values.

File: packages/openstef-models/tests/unit/transforms/general/test_imputer.py [hunk 2]
```diff
@@ -34,23 +41,88 @@ def sample_dataset() -> TimeSeriesDataset:
...
     # Arrange & Act & Assert
     with pytest.raises(ValidationError, match="fill_value must be provided when imputation_strategy is CONSTANT"):
         Imputer(imputation_strategy="constant")
 
 
-def test_basic_imputation_works(sample_dataset: TimeSeriesDataset):
-    """Test that basic imputation removes NaN values."""
-    # Arrange
-    transform = Imputer(imputation_strategy="mean")
+@pytest.mark.parametrize(
```
Comment (fleur-petit) [inline packages/openstef-models/tests/unit/transforms/general/test_imputer.py:71]: added parametrize to test if basic imputation works for all strategies.

File: packages/openstef-models/tests/unit/transforms/general/test_imputer.py [hunk 3]
```diff
@@ -225,3 +297,99 @@ def test_no_missing_values_data_preservation():
...
+
+    mean_error_temp = abs(mean_temperature - true_temperature)
+    iter_error_temp = abs(iter_temperature - true_temperature)
+
+    # Iterative should be closer to true values
+    assert iter_error_rad < mean_error_rad
+    assert iter_error_temp < mean_error_temp
+
+
+def test_iterative_imputer_only_one_feature():
```
Comment (fleur-petit) [inline packages/openstef-models/tests/unit/transforms/general/test_imputer.py:349]: @bartpleiter is it clear to you what this test does?

File: packages/openstef-models/tests/unit/transforms/general/test_imputer.py [hunk 4]
```diff
@@ -225,3 +297,99 @@ def test_no_missing_values_data_preservation():
...
+        initial_strategy="median",
+        max_iterations=5,
+        tolerance=1e-3,
+    )
+
+    # Act
+    transform.fit(dataset)
+
+    # Assert
+    with pytest.warns(UserWarning, match="Iterative imputer with only one feature"):
```
Comment (fleur-petit) [inline packages/openstef-models/tests/unit/transforms/general/test_imputer.py]: @bartpleiter do you think this warning is useful?

File: packages/openstef-models/src/openstef_models/transforms/general/imputer.py [hunk 5]
```diff
@@ -98,6 +110,32 @@ class Imputer(BaseConfig, TimeSeriesTransform):
...
+    ... )
+    >>> transform_iterative.fit(dataset)
+    >>> result_iterative = transform_iterative.transform(dataset)
+    >>> result_iterative.data["temperature"].isna().sum() == 0  # Temperature NaNs filled
+    np.True_
+    >>> np.isnan(result_iterative.data["radiation"][1])  # Radiation first NaN replaced
+    np.False_
+    >>> np.isnan(result_iterative.data["radiation"][3]) # Check if trailing NaN is preserved
+    np.True_
+    >>> result_iterative.data["wind_speed"].isna().sum() == 2  # Check if trailing NaNs are preserved
```
Comment (bartpleiter) [inline packages/openstef-models/src/openstef_models/transforms/general/imputer.py]:
```suggestion
    >>> result_iterative.data["wind_speed"].isna().sum() == 2  # Wind speed NaNs preserved
```

File: packages/openstef-models/src/openstef_models/transforms/general/imputer.py (hunk 1 above)
Comment (bartpleiter) [reply packages/openstef-models/src/openstef_models/transforms/general/imputer.py:171]: I think this is a good place to indicate that there is no way to differentiate between the selected columns for imputation and the predictor columns for imputation. Explanation is clear, I would maybe rephrase it a bit into: 

"Features to impute. If strategy is 'iterative', then these features are also used as predictors for the imputation."

File: packages/openstef-models/src/openstef_models/transforms/general/imputer.py [hunk 6]
```diff
@@ -174,6 +240,15 @@ def transform(self, data: TimeSeriesDataset) -> TimeSeriesDataset:
...
 
+        # Warn if using iterative imputation with only one feature
+        if self.imputation_strategy == "iterative" and len(features) < 2:  # noqa: PLR2004
+            warnings.warn(
+                "Iterative imputer with only one feature will fall back to initial_strategy. "
+                f"Using '{self.initial_strategy}' for imputation.",
+                UserWarning,
+                stacklevel=2,
+            )
+
```
Comment (bartpleiter) [inline packages/openstef-models/src/openstef_models/transforms/general/imputer.py:250]: Nice to warn the user on this case! Let's use the logger for this though instead of warnings. See https://github.com/OpenSTEF/openstef/blob/b6c13fa3994bb86f6114236a14c70fe224468e0d/packages/openstef-models/src/openstef_models/models/forecasting_model.py#L291-L295 for an example

File: packages/openstef-models/tests/unit/transforms/general/test_imputer.py [hunk 7]
```diff
@@ -34,23 +41,88 @@ def sample_dataset() -> TimeSeriesDataset:
...
+    """Create sample dataset with simple linear relation between features where temperature = 2 * radiation.
+
+    Returns:
+        A TimeSeriesDataset with 2 features: radiation and temperature (with some NaN),
+        spanning 5 hours.
+    """
+    data = pd.DataFrame(
+        {
+            "radiation": [10.0, np.nan, 30.0, 40.0, 50.0],
+            "temperature": [20.0, 40.0, 60.0, np.nan, 100.0],  # Missing: should be ~80
```
Comment (bartpleiter) [inline packages/openstef-models/tests/unit/transforms/general/test_imputer.py]: Let's remove this comment as I think it is not needed (and tells something about what will be done with the data), and radiation also does not have it
```suggestion
            "temperature": [20.0, 40.0, 60.0, np.nan, 100.0],
```

File: packages/openstef-models/tests/unit/transforms/general/test_imputer.py (hunk 2 above)
Comment (bartpleiter) [reply packages/openstef-models/tests/unit/transforms/general/test_imputer.py:71]: Great to use parametrize! However, now the test becomes quite complex to read/maintain due to both parameters and the nested if/elif structure within the test. Perhaps it might be better to split the test up into two, where one test tests the non-iterative strategies, and the other one tests the iterative ones.

File: packages/openstef-models/tests/unit/transforms/general/test_imputer.py [hunk 8]
```diff
@@ -225,3 +297,99 @@ def test_no_missing_values_data_preservation():
...
+
+    # Act
+    transform.fit(correlated_dataset)
+    result = transform.transform(correlated_dataset)
+
+    # Assert
+    imputed_value = cast(float, result.data.loc[result.data.index[3], "temperature"])
+    expected_value = cast(float, correlated_dataset.data.loc[correlated_dataset.data.index[3], "radiation"]) * 2
+    # Allow some tolerance since it's ML-based
+    assert abs(imputed_value - expected_value) < 10.0, f"Expected ~{expected_value}, got {imputed_value}"
```
Comment (bartpleiter) [inline packages/openstef-models/tests/unit/transforms/general/test_imputer.py]: I think it would be nice to also add a check for the imputed value of radiation[1] (as it should also be getting imputed, right?). Might be good to slightly reduce the tolerance though. Given the size of the test dataset a few percent instead of ~10 percent should be fast enough.

File: packages/openstef-models/tests/unit/transforms/general/test_imputer.py [hunk 9]
```diff
@@ -225,3 +297,99 @@ def test_no_missing_values_data_preservation():
...
+
+
+def test_iterative_imputer_better_than_mean(correlated_dataset: TimeSeriesDataset):
+    """Test that iterative imputation outperforms mean imputation for linearly related features.
+
+    When there is a simple linear relationship, and the missing values are not coincidentally the mean, iterative
+    imputer should do better than mean imputer.
+    """
```
Comment (bartpleiter) [inline packages/openstef-models/tests/unit/transforms/general/test_imputer.py]: I do not fully understand this test 😅. Given the previous test with the expected values, does this test still add value? (e.g. could it fail if the previous test passes?)

File: packages/openstef-models/tests/unit/transforms/general/test_imputer.py (hunk 3 above)
Comment (bartpleiter) [reply packages/openstef-models/tests/unit/transforms/general/test_imputer.py:349]: I think I do: it tests whether the warning is raised when only one feature is used, and makes sure there is output based on the initial strategy.

File: packages/openstef-models/tests/unit/transforms/general/test_imputer.py (hunk 4 above)
Comment (bartpleiter) [reply packages/openstef-models/tests/unit/transforms/general/test_imputer.py]: Yes I do (as a logging warning). You probably have to adjust this assert to check if a warning has been thrown by the logger

File: packages/openstef-models/tests/unit/transforms/general/test_imputer.py (hunk 9 above)
Comment (fleur-petit) [reply packages/openstef-models/tests/unit/transforms/general/test_imputer.py]: good point, will remove it.

File: packages/openstef-models/tests/unit/transforms/general/test_imputer.py (hunk 8 above)
Comment (fleur-petit) [reply packages/openstef-models/tests/unit/transforms/general/test_imputer.py]: good point, will add it to assert.

**Review by bartpleiter — APPROVED** [review]
Looks good to me!
